parla/task_runtime.py

```diff
@@ -1255,23 +1254,54 @@ def run_mapper(self):
                 # move to the mapped task scheduling.
                 break
 
-    def run_deviceq_mapper(self):
+    def _schedule_tasks(self):
+        """ Currently this doesn't do any intelligent scheduling (ordering).
+            Dequeue all ready tasks and send them to device queues in order.
+        """
+        print("[Scheduler] Schedule Phase")
         while True:
             task: Optional[TaskBase] = self._dequeue_task()
             if not task or not task.assigned:
                 logger.debug("Task %r: Failed to assign", task)
                 break
             for d in task.req.devices:
-                print("[Scheduler] Task, ", task.name, ", is enqueued onto a device Q.", sep='')
-                worker = self._worker_threads[task.req.env_no]
-                worker._enqueue_task_local(task)
+                print(f"[Scheduler] Enqueuing {task.name} to device {d}")
+                self._device_queues[d].append(task)
+
+    def _launch_tasks(self):
+        """ Iterate through free devices and launch tasks on them
+        """
+        print("[Scheduler] Launch Phase")
+        with self._monitor:
+            for dev, queue in self._device_queues.items():
+                # Make sure there's an available WorkerThread
+                if len(self._free_worker_threads) == 0:
+                    break
+
+                # Hack to make launching work for now
+                # Only works for single-device tasks, and only one task at a time on each device
+                # TODO: Make this not terrible
+                if self._available_resources._occupancy_dict[dev] == False: # if not occupied
+                    if len(queue) > 0: # if there are tasks on the queue
+                        task = queue.pop() # grab a task
+                        worker = self._free_worker_threads.pop() # grab a worker
+                        print(f"[Scheduler] Launching {task.name} on WorkerThread {worker.index}")
+                        self._available_resources._occupancy_dict[dev] = True # mark the device as occupied
+                        worker.assign_task(task) # assign the task to the worker (this notifies the worker's monitor)
+                        print(f"[Scheduler] Launched {worker.task.name}")
```

Comment:
`_launch_tasks` is the main new thing in this commit and all the other changes are there to just make this work.

Reply:
this looks good!